### ISSUE TITLE
Long delays in repeated prisma:engine:connection (reconnects excessively)

### ISSUE THREAD

**Issue by kitten** [prisma/prisma#20088]
### Bug description

Our API runs Prisma on Fly.io connecting as a MySQL client to a database on Planetscale.
This is a GraphQL API and we have a particular query that, without Redis caching, can have a depth of about 4–6 parallel batches of queries.

This was working fine locally, as expected (connecting to either MariaDB or MySQL), and we debugged some of the queries and minimised time spent on them by optimising some indexes.

In production, however, we observed that the time spent on the query wasn't decreasing and that there was a random but high cost associated to each batch of resolvers.

After enabling `tracing` and sending the time slices to Sentry via OpenTelemetry, we were able to observe that while the first connection Prisma uses (as indicated by `prisma:engine:connection`) may be “established” `0.02ms`, i.e. it's reused.

However, subsequent `prisma:engine:connection` events pop up that are indicated to take between `170ms` to `230ms`. This is repeated on each resolver level (potentially even multiple per resolver level) and they add up quickly.

For context, most `prisma:engine:db_query` events are about `60ms` long. (This number is pretty consistent).

### What we've tried so far

- Before, we used to have resolver events sent to Sentry
  - Naturally, hese just indicate that some resolvers suddenly take up much more time than expected, since they can't see Prisma events granularly
- When logging `prisma`'s `query` event we did see `SELECT 1`s pop up during queries
  - This did indicate to us that Prisma correctly attempts to keep the connection alive
- The total roundtrip time from the Fly machine to the DB is up to `25ms` on a new connection
  - This is based on running a mock HTTP request and timing it via `curl`
  - Both Fly and the Planetscale instance are basically close together in US East
- Planetscale itself reports that _all_ database queries related to this finishin in under `6ms`
- Altering `connection_limit` has not changed this issue
  - The time and reconnections are also consistent over many requests after a restart well beyond multiple times the connection limit
- The Docker environment doesn't seem to be part of the issue
  - We've switched from `node:18-bullseye-slim` to node:18-alpine` and it didn't affect the issue
  - i.e. `debian-openssl-1.1.x` and `linux-musl-openssl-3.0.x` of the `libquery` engine were tested and both observed the same behaviour
- We used `mysql2` on a healthcheck endpoint to test against Prisma's times
  - A query takes `10ms-30ms` on an established connection
  - A query while establishing a new connection takes `40-60ms`
  - Planetscale's serverless driver typically can beat this time, but `mysql2` never does worse than the above
- I believe, we tested `binary` rather than `library` mode as well
- We also downgraded Prisma a couple of versions to test (which we currently couldn't due to features used), but this didn't affect the issue

<details>
<summary>A screenshot of a Prisma trace from our Sentry</summary>

![Screenshot 2023-07-06 at 11-59-08 Trace Details — Performance — speakeasy-ai — Sentry](https://github.com/prisma/prisma/assets/2041385/e95a5734-3ee1-4e2a-acb5-66e8478837af)

</details>

### Expected behavior

- `prisma:engine:connection` should not take up any time once the pool has established all connections
- or: `prisma:engine:connection` is not expected to take up `>60ms` comparatively when it does happen

### Connection and Generator part of `schema.prisma`

- `DATABASE_URL` vaguely looks like: `mysql://[...]:pscale_pw_[...]@aws.connect.psdb.cloud/everywhere?sslaccept=strict`

```prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["views", "jsonProtocol", "tracing"]
  output          = "../generated/prisma-client"
}

generator pothos {
  provider = "prisma-pothos-types"
}

datasource db {
  provider     = "mysql"
  url          = env("DATABASE_URL")
  relationMode = "prisma"
}
```

### How we're creating the Prisma client

- `NODE_ENV` is set to `"production"` when deployed
- `EXPLAIN` is set to `false`

```ts
import { Prisma, PrismaClient } from "@everywhere/db";

const db: PrismaClient = new PrismaClient({
  log: [
    process.env.NODE_ENV !== "production" && {
      emit: process.env.EXPLAIN ? "event" : "stdout",
      level: "query",
    },
    process.env.NODE_ENV !== "production" && {
      emit: "stdout",
      level: "warn",
    },
    {
      emit: "stdout",
      level: "error",
    },
  ].filter((x): x is Prisma.LogDefinition => !!x),
});

if (process.env.EXPLAIN) {
  (db.$on as any)("query", async (event: Prisma.QueryEvent) => {
    if (!/^\s*SELECT/.test(event.query)) return;

    const result = await db.$queryRawUnsafe(
      `EXPLAIN ${event.query}`,
      ...JSON.parse(event.params)
    );

    console.log("Query: ", event.query);
    console.log("Duration: ", event.duration + "ms");
    console.log("Params: ", event.params);
    console.log("Explain: ", result);
  });
}

export * from "@everywhere/db";
export default db;
```

### Environment & setup

- OS: Docker Alpine Linux
- Database: PlanetScale
- Node.js version: v18.16.1

### Prisma Version

**NOTE:** Only the version is representative of the production environment.

```
prisma                  : 4.16.1
@prisma/client          : 4.16.1
Current platform        : darwin-arm64
Query Engine (Node-API) : libquery-engine b20ead4d3ab9e78ac112966e242ded703f4a052c (at node_modules/.pnpm/@prisma+engines@4.16.1/node_modules/@prisma/engines/libquery_engine-darwin-arm64.dylib.node)
Migration Engine        : migration-engine-cli b20ead4d3ab9e78ac112966e242ded703f4a052c (at node_modules/.pnpm/@prisma+engines@4.16.1/node_modules/@prisma/engines/migration-engine-darwin-arm64)
Format Wasm             : @prisma/prisma-fmt-wasm 4.16.0-66.b20ead4d3ab9e78ac112966e242ded703f4a052c
Default Engines Hash    : b20ead4d3ab9e78ac112966e242ded703f4a052c
Studio                  : 0.484.0
```

**Comment by janpio** [comment]
How many connections are available to Prisma (via `connection_limit`)? How many queries are being run in parallel (= at the same time) in these situations? Can you observe the connection creation on the database side?

**Comment by kitten** [comment]
> How many connections are available to Prisma (via connection_limit)?

We've tried a couple of numbers — anywhere between `3` and `50`.
Currently, we've set it back to the default, which should be `3` anyway.

> How many queries are being run in parallel (= at the same time) in these situations?

This is harder for me to tell, since Sentry isn't showing it that well, and I haven't checked locally in a while, but if I had to guess it's probably 6–9.

> Can you observe the connection creation on the database side?

We can't in production. Planetscale doesn't give us metrics on the number of connections established over time it seems.

> How does your API execute the multiple queries it needs to fulfill 1 request?

I'm not sure I understand the question exactly. We used to use Pothos’ composition of queries, which meant that we had a couple of queries that were blocking sub-resolver queries.
A lot of these in the hotpath and all in the query the screenshot is from (which is the most expensive GraphQL query) have now been replaced with dataloader-based queries.

So, several queries may be sent in parallel per level of resolvers. The next level of resolvers will then send the next batch of queries, i.e. multiple Prisma client calls, and so on.

**Comment by janpio** [comment]
> We've tried a couple of numbers — anywhere between `3` and `50`. Currently, we've set it back to the default, which should be `3` anyway.
> 
> ...
> 
> This is harder for me to tell, since Sentry isn't showing it that well, and I haven't checked locally in a while, but if I had to guess it's probably 6–9.

This would explain some queries having to wait to get a connection: Prisma has 3 connections to the database available, but is asked to run 6 queries at the same time for 1 request to your app. It will open 3 connections (if they do not exist), and then handle the first 3 queries with them. In the meantime the other 3 queries need to wait until the responses come in for these queries and the connection is returned to the connection pool. Afterwards the other 3 queries can get their own connection and be executed.

Multiple requests at the same time add on top of that. (If the queue gets really long, the queries might not even be handled in the `pool_timeout` duration, which defaults to 10s. You will get a special error in these cases.)

> > How does your API execute the multiple queries it needs to fulfill 1 request?
> 
> I'm not sure I understand the question exactly. We used to use Pothos’ composition of queries, which meant that we had a couple of queries that were blocking sub-resolver queries. A lot of these in the hotpath and all in the query the screenshot is from (which is the most expensive GraphQL query) have now been replaced with dataloader-based queries.
> 
> So, several queries may be sent in parallel per level of resolvers. The next level of resolvers will then send the next batch of queries, i.e. multiple Prisma client calls, and so on.

The way your application "spawns" queries can have an influence. Does it actually do a `Promise.all` under the hood for example? That would be the most extreme variant.

In these problems it is usually very useful to really understand how the app works and triggers queries. I would probably create a local setup where I can really monitor each incoming connection and query to the database while I send 1 request (which optimally will have a lot more loggin in the app so I can also really understand what is being triggered there.)

**Comment by janpio** [comment]
> Clearly, that can't be the case?

Exactly that is the case unfortunately. 1 query per 1 connection at a time. That is why it is so easy to run out of pool connections when queries get slower - as you wait for the return from the database, the connection is blocked and other queries have to wait - or when you get more traffic that causes more requets on your application - one server with an application that has x connections can only ever run x queries at a time.

> If that's the case however, and as we in the past raised the limit to 50 and then ran queries to fill up the pool, how would this not lead to the expected behaviour and connections would still die / be re-established?

That is the correct question. Did the 50 connections actually actively get used? If your system is currently only queueing queries for a short amount of times (e.g. 1 other query has to finish before the new one can get a connection), then doubling or 10x-ing the pool size (`connection_limit`)  obviously should trivially resolve this. So do you have any evidence if Prisma did not even try to use these available connections? If it did use them, were there still more queries that had to wait for a connection?

**Comment by kitten** [comment]
Just to leave an update here; We did speculate that this may be down to a problem with concurrency in running the queries, but initially dismissed it as unlikely, since it'd be quite the problem in scaling long-running machines that handle many connections, if that's the case.

So, what we'll try is to scale up the `connection_limit` back to `50`, increase the `pool_timeout`, if applicable, and scale the Fly VMs up to dual-cores from single-core machines, just to ensure it's not a threading issue in Prisma.

**Comment by kitten** [comment]
Hiya 👋 
Sorry for the late update.

Basically, we increased the `connection_limit` and ensured that all connections are established across all parallel machines before checking the stats. However, unfortunately, we still see sporadic `prisma:engine:connection` events in our traces.
This has become rarer, but it's still at a likelihood where we're seeing them for almost every request.

For now, since we evaluated that going with something lower level that simply generates SQL queries directly is faster, we probably can't provide more data or information here.
Basically, we evaluated that we aren't equipped to analyse when `prisma:engine:connection` occurs, why it takes relatively long, and why `prisma:engine:db_query` (even without serialize timings/events) is taking longer compared to `mysql2` (given that `@planetscale/database` wouldn't be a fair comparison)

So, I'm happy to close this, but to summarise this, I'm still convinced something isn't quite right here. 😅

**Comment by janpio** [comment]
Then let's keep this open, and we will take a look when we have the capacity to do so. Thanks for all the information you provided @kitten.

**Comment by CapCap** [comment]
![image](https://github.com/prisma/prisma/assets/1482859/077932b8-047b-4197-ae92-27ff959fc4e5)
we're seeing the same thing- alloydb (PG) on google cloud. 
Prisma client/instrumentation version: 5.3.1
